src/java/org/apache/cassandra/security/PEMReader.java

```diff
@@ -0,0 +1,249 @@
+
+        if (keyPassword != null)
+        {
+            logger.debug("Encrypted key's length: {}",derKeyBytes.length);
+            logger.debug("Key's password length: {}", keyPassword.length());
+
+            EncryptedPrivateKeyInfo epki = new EncryptedPrivateKeyInfo(derKeyBytes);
+            logger.debug("Encrypted private key info's algorithm name: {}", epki.getAlgName());
+
+            AlgorithmParameters params = epki.getAlgParameters();
+            PBEKeySpec pbeKeySpec = new PBEKeySpec(keyPassword.toCharArray());
+            Key encryptionKey = SecretKeyFactory.getInstance(epki.getAlgName()).generateSecret(pbeKeySpec);
+            pbeKeySpec.clearPassword();
+            logger.debug("Key algorithm: {}", encryptionKey.getAlgorithm());
+            logger.debug("Key format: {}", encryptionKey.getFormat());
+
+            Cipher cipher = Cipher.getInstance(epki.getAlgName());
+            cipher.init(Cipher.DECRYPT_MODE, encryptionKey, params);
+            byte[] rawKeyBytes = cipher.doFinal(epki.getEncryptedData());
+            logger.debug("Decrypted private key's length: {}", rawKeyBytes.length);
+
+            keySpec = new PKCS8EncodedKeySpec(rawKeyBytes);
+        }
+        else
+        {
+            logger.debug("Key length: {}",derKeyBytes.length);
+            keySpec = new PKCS8EncodedKeySpec(derKeyBytes);
+        }
+
+        PrivateKey privateKey = null;
+
+        /*
+         * Ideally we can inspect the OID (Object Identifier) from the private key with ASN.1 parser and identify the
+         * actual algorithm of the private key. For doing that, we have to use some special library like BouncyCastle.
+         * However in the absence of that, below brute-force approach can work- that is to try out all the supported
+         * private key algorithms given that there are only three major algorithms to verify against.
+         */
+        for(int i=0;i<SUPPORTED_PRIVATE_KEY_ALGORITHMS.length;i++)
+        {
+            try
+            {
+                privateKey = KeyFactory.getInstance(SUPPORTED_PRIVATE_KEY_ALGORITHMS[i]).generatePrivate(keySpec);
+                logger.info("Parsing for the private key finished with {} algorithm.",
+                            SUPPORTED_PRIVATE_KEY_ALGORITHMS[i]);
+                return privateKey;
+            }
+            catch(Exception e)
+            {
+                logger.debug("Failed to parse the private key with {} algorithm. Will try the other supported " +
+                             "algorithms.", SUPPORTED_PRIVATE_KEY_ALGORITHMS[i]);
+            }
+        }
+
+       if(privateKey==null)
+       {
+           throw new GeneralSecurityException("The given private key could not be parsed with any of the supported " +
+                                              "algorithms. Please see PEMReader#SUPPORTED_PRIVATE_KEY_ALGORITHMS.");
+       }
+       // Must never come here
+       return null;
```

Comment:
Why the test on `privateKey` - could just always `throw` then you would never return `null`.

Reply:
Good catch!